Source/WebCore/dom/FindRevealAlgorithms.cpp

```diff
@@ -37,49 +37,66 @@
 
 namespace WebCore {
 
-// https://html.spec.whatwg.org/#ancestor-details-revealing-algorithm
-void revealClosedDetailsAncestors(Node& node)
+enum class RevealType : bool {
+    ClosedDetails,
+    HiddenUntilFound
+};
+
+// https://html.spec.whatwg.org/#ancestor-revealing-algorithm
+void revealClosedDetailsAndHiddenUntilFoundAncestors(Node& node)
 {
-    if (!node.document().settings().detailsAutoExpandEnabled())
+    node.protectedDocument()->updateStyleIfNeeded();
+
+    // Bail out if there is neither a hidden=until-found or details ancestor.
+    if (node.renderStyle() && !node.renderStyle()->autoRevealsWhenFound())
         return;
 
-    Ref currentNode = node;
-    while (RefPtr parent = currentNode->parentInComposedTree()) {
-        if (RefPtr slot = currentNode->assignedSlot(); slot && slot->userAgentPart() == UserAgentParts::detailsContent() && slot->shadowHost()) {
-            currentNode = *slot->shadowHost();
-            Ref details = downcast<HTMLDetailsElement>(currentNode);
+    auto closedDetailsElementAncestor = [](Node& node) -> RefPtr<HTMLDetailsElement> {
+        RefPtr slot = node.assignedSlot();
+        if (slot && slot->userAgentPart() == UserAgentParts::detailsContent() && slot->shadowHost()) {
+            Ref details = downcast<HTMLDetailsElement>(*slot->shadowHost());
             if (!details->hasAttributeWithoutSynchronization(HTMLNames::openAttr))
-                details->toggleOpen();
-        } else
-            currentNode = *parent;
-    }
-}
+                return details;
+        }
+        return nullptr;
+    };
 
-// https://html.spec.whatwg.org/#ancestor-hidden-until-found-revealing-algorithm
-void revealHiddenUntilFoundAncestors(Node& node)
-{
-    if (!node.document().settings().hiddenUntilFoundEnabled())
-        return;
+    Vector<std::pair<Ref<HTMLElement>, RevealType>> ancestors;
+
+    RefPtr ancestor = node;
+    while (ancestor->parentElementInComposedTree()) {
+        if (RefPtr element = dynamicDowncast<HTMLElement>(*ancestor); element && element->isHiddenUntilFound())
+            ancestors.append({ *element, RevealType::HiddenUntilFound });
+        if (RefPtr details = closedDetailsElementAncestor(*ancestor)) {
+            ancestors.append({ *details, RevealType::ClosedDetails });
+            ancestor = *details;
```

Comment:
Seems like this could fall through so it moves to the parent instead of checking the same node a second time?

Reply:
This would break the case where details has "hidden=until-found" as well, where we need to check the details element for that attribute. The details element can also be nested in another details, so we'd need to make sure that's also being checked. This is covered by some of the tests.

Reply:
I've dropped the `ancestor = *details;`, it was supposed to be a micro perf optimization, but it's making the code a bit harder to read (and does not match the spec)